csrc/runtime/executor.cpp

```diff
@@ -1035,29 +1160,37 @@ KernelArgumentHolder KernelExecutor::run(
         }
       }
       if (has_expansion) {
-        intermediate_buffer =
-            at::native::expand(intermediate_buffer, buf_info.sizes);
+        intermediate_buffer = at::native::expand(
+            intermediate_buffer, buf_info.shape_info.allocation_sizes);
       }
       args.push(intermediate_buffer);
       intermediate_args.push(intermediate_buffer);
-      expr_eval.bind(
-          compiled_kernel_->kernel()
-              ->summary()
-              .global_allocations.at(intermediate_i)
-              ->buffer(),
-          args
-              [compiled_kernel_->kernel()->inputs().size() +
-               output_args.size() + intermediate_i]);
       if (buf_info.is_profile_buffer) {
         profile_buffer = intermediate_buffer;
       }
     }
   }
 
-  if (executor_entry->args.empty()) {
-    computeArgs(*executor_entry, expr_eval, compiled_kernel_->kernel());
+  if (args.size() != compiled_kernel_->kernel()->parameters().size()) {
+    NVF_ERROR(
+        has_tma_ || has_rng_,
+        "No TMA or RNG found in the kernel, but detected an argument size mismatch.");
+    // If args don't match one of two things is happening. We need to add TMA
+    // related args or RNG related args. Resolve these scenarios.
+    if (has_tma_) {
+      // Resolving TMA requires binding all values and evaluating the TMA
+      // arguments
+      args = resolveTMA(*executor_entry, args);
+    } else if (has_rng_) {
```

Comment:
So we an not use TMA and RNG in the same kernel?

Reply:
resolveTMA will resolve RNG values too. resolveTMA actually resolves just about any argument that's missing values

Reply:
In the future I'd like to cleanup the resolution of missing params to only bind necessary inputs for their evaluation. That would help unify the different runtime resolutions necessary (dynamic alias, rng, tma, and anything else we come up with). I'd similarly like to do something like this for expression evaluator executor/host ir.

Reply:
This makes sense. For now, could you update the comment in code saying that this also resolves RNG? Otherwise this looks like a bug.

Reply:
Will do.

Reply:
Added